### PR TITLE
Promote Arcade builds using Darc

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/4677

Use Darc to promote Arcade builds to the final channel.

### CODE REVIEW

**Comment by riarenas** [comment]
1. > A failed execution will look like this

    Does that mean if this happens the build will fail?

2. How does the error bubble up if the build promotion pipeline encounters an error?

**Comment by JohnTortugo** [comment]
> * Does that mean if this happens the build will fail?
> *

For those case there is no build as it was a validation (or a failure) in Darc `add-build-to-channel`.

> How does the error bubble up if the build promotion pipeline encounters an error?

There is no bubble up error reporting. Darc triggers the build asynchronously.

**Comment by JohnTortugo** [comment]
Pushed some changes.

- The script won't fail the build when the build being promoted is already in the intended channel.
- The script will wait until the promotion build finishes execution and then it'll reports the result.
- Had to add a new parameter to the script to receive the AzDO PAT

Note that all these new changes will be ported to Darc; @riarenas and I talked offline and decided to implement them here so we don't need to wait for a new Darc deployment before merging this PR.

**Comment by riarenas** [comment]
Is there an issue for the work of giving `darc add-build-to-channel` the ability to wait for completion?

File: eng/validation/update-channel.ps1 [hunk 1]
```diff
@@ -14,56 +15,72 @@ function Get-Headers([string]$accept, [string]$barToken) {
...
+        exit 1
+	}
 
-$getChannelsEndpoint = "$maestroEndpoint/api/channels?api-version=2018-07-16"
-$channels = Invoke-WebRequest -Uri $getChannelsEndpoint -Headers $headers | ConvertFrom-Json
-$matchedChannel = $($channels | Where-Object -Property "name" -Value "${targetChannelName}" -EQ | Select-Object -Property id)
+    if ($DarcOutput -match "has already been assigned to") 
+    {
+        Write-Host "Build has already been assigned to channel '${targetChannelName}'."
+	}
```
Comment (riarenas) [inline eng/validation/update-channel.ps1]: [nit] weird indenting

Comment (riarenas) [reply eng/validation/update-channel.ps1]: Here and in the else.

**Comment by JohnTortugo** [comment]
> Is there an issue for the work of giving `darc add-build-to-channel` the ability to wait for completion?

Here: https://github.com/dotnet/arcade/issues/4863

File: eng/validation/update-channel.ps1 [hunk 2]
```diff
@@ -14,56 +15,72 @@ function Get-Headers([string]$accept, [string]$barToken) {
...
+        Write-Host $DarcOutput
+        exit 1
+	}
 
-$getChannelsEndpoint = "$maestroEndpoint/api/channels?api-version=2018-07-16"
-$channels = Invoke-WebRequest -Uri $getChannelsEndpoint -Headers $headers | ConvertFrom-Json
-$matchedChannel = $($channels | Where-Object -Property "name" -Value "${targetChannelName}" -EQ | Select-Object -Property id)
+    if ($DarcOutput -match "has already been assigned to") 
+    {
+        Write-Host "Build has already been assigned to channel '${targetChannelName}'."
```
Comment (riarenas) [inline eng/validation/update-channel.ps1]: The old message had a little more information. 

`Write-Host "Build '$buildId' is already in channel '$targetChannelName'. This is most likely an arcade-validation internal build`

File: eng/validation/update-channel.ps1 [hunk 3]
```diff
@@ -14,56 +15,72 @@ function Get-Headers([string]$accept, [string]$barToken) {
...
-    Write-Host "POSTing to $postBuildIntoChannelApiEndpoint..."
-    $postResponse = Invoke-WebRequest -Uri $postBuildIntoChannelApiEndpoint -Headers $headers -Method Post
-    Write-Host "Build '$buildId' was successfully added to channel '$channelId'"
+            Start-Sleep -Seconds $waitIntervalsInSeconds
+
+            $build = Get-AzDO-Build -token $azdoToken -azdoBuildId $azdoBuildId
+	    } while ($build.status -ne "completed")
+
+        if ($build.result -eq "succeeded") {
+            Write-Host "The build was promoted with success."
```
Comment (riarenas) [inline eng/validation/update-channel.ps1]: I liked the old message better: `"Build '$buildId' was successfully added to channel '$channelId'"`

**Comment by riarenas** [comment]
How did you validate this?

**Comment by JohnTortugo** [comment]
> How did you validate this?

Locally, by calling the script on the command line.

File: eng/validation/update-channel.ps1 [hunk 4]
```diff
@@ -14,56 +15,72 @@ function Get-Headers([string]$accept, [string]$barToken) {
...
+        Write-Host "Problems using Darc to promote build ${buildId} to channel ${targetChannelName}. Stopping execution..."
+        Write-Host $DarcOutput
+        exit 1
+	}
 
-$getChannelsEndpoint = "$maestroEndpoint/api/channels?api-version=2018-07-16"
-$channels = Invoke-WebRequest -Uri $getChannelsEndpoint -Headers $headers | ConvertFrom-Json
-$matchedChannel = $($channels | Where-Object -Property "name" -Value "${targetChannelName}" -EQ | Select-Object -Property id)
+    if ($DarcOutput -match "has already been assigned to") 
+    {
```
Comment (riarenas) [inline eng/validation/update-channel.ps1]: [nit] The formatting of this whole block still seems a little off. The if and else don't match the format for the rest of the file for opening and closing braces.

File: eng/validation/update-channel.ps1 [hunk 5]
```diff
@@ -14,56 +15,74 @@ function Get-Headers([string]$accept, [string]$barToken) {
...
 
-if (!$matchedChannel) {
-    Write-Host "Channel with name '$targetChannelName' was not found..."
-    exit 1    
-}
+    if ($DarcOutput -match "has already been assigned to") {
+        Write-Host "Build '$buildId' is already in channel '$targetChannelName'. This is most likely an arcade-validation internal build"
+    }
+    else {
+        $buildUrlRegex = "https://dnceng.visualstudio.com/internal/_build/results\?buildId=(?<buildId>[0-9]*)"
```
Comment (riarenas) [inline eng/validation/update-channel.ps1]: [nit] The indent of this block still seems off

Comment (JohnTortugo) [reply eng/validation/update-channel.ps1]: It's actually aligned :-| . 8 spaces.

**Comment by riarenas** [comment]
@JohnTortugo 
This ended up causing a failure: 
https://dev.azure.com/dnceng/internal/_build/results?buildId=526341

I missed the fact that we'll need to pass the BAR token + possibly the azdo PAT to the darc call.

**Comment by JohnTortugo** [comment]
I'm taking a look now.

**Pull request merged**